File: src/parser/parser.ts
import { artefactKeys, nodeKeywords, type CairnAst, type CairnNode, type Edge, type NodeKind } from "./ast.js";
import { lex, type Token, type TokenType } from "./lexer.js";
import { ParseError } from "./errors.js";

const validNodeKeywords = nodeKeywords.join(", ");

export function parseCairnDsl(input: string, source = "<input>"): CairnAst {
  return new Parser(lex(input, source), source).parse();
}

class Parser {
  private index = 0;

  constructor(
    private readonly tokens: Token[],
    private readonly source: string,
  ) {}

  parse(): CairnAst {
    const root = this.parseNode(["System"]);
    const edges: Edge[] = [];

    while (!this.check("eof")) {
      if (this.check("identifier") && this.checkNext("arrow")) {
        edges.push(this.parseEdge());
        continue;
      }

      const token = this.peek();
      if (this.check("identifier")) {
        throw new ParseError(`Unknown keyword '${token.value}'. Valid declarations: ${validNodeKeywords}`, {
          source: this.source,
          position: token.position,
        });
      }
      throw this.expected("edge or end of file");
    }

    return { root, edges };
  }

  private parseNode(allowedKinds: NodeKind[]): CairnNode {
    const keyword = this.consume("identifier", "node keyword");
    if (!nodeKeywords.includes(keyword.value as NodeKind)) {
      throw new ParseError(`Unknown keyword '${keyword.value}'. Valid declarations: ${validNodeKeywords}`, {
        source: this.source,
        position: keyword.position,
      });
    }
    if (!allowedKinds.includes(keyword.value as NodeKind)) {
      throw new ParseError(`${keyword.value} is not valid in this position`, {
        source: this.source,
        position: keyword.position,
      });
    }

    const name = this.consume("identifier", "node name");
    const description = this.consume("string", "node description");
    const node: CairnNode = {
      kind: keyword.value as NodeKind,
      name: name.value,
      description: description.value,
      tags: [],
      paths: [],
      artefacts: {},
      children: [],
      position: keyword.position,
    };

    this.parseInlineNodeMetadata(node);

    if (this.match("braceOpen")) {
      while (!this.check("braceClose") && !this.check("eof")) {
        if (this.check("identifier") && nodeKeywords.includes(this.peek().value as NodeKind)) {
          node.children.push(this.parseNode(["Container", "Module", "Actor"]));
        } else {
          this.parseNodeAttribute(node);
        }
      }
      this.consume("braceClose", "}");
    }

    return node;
  }

  private parseInlineNodeMetadata(node: CairnNode): void {
    while (!this.check("braceOpen") && !this.check("braceClose") && !this.check("eof")) {
      if (this.check("tag")) {
        node.tags.push(this.advance().value);
        continue;
      }
      if (this.check("identifier") && this.peek().value === "id") {
        this.advance();
        node.id = this.consume("string", "id value").value;
        continue;
      }
      break;
    }
  }

  private parseNodeAttribute(node: CairnNode): void {
    const key = this.consume("identifier", "attribute name");
    if (key.value === "id") {
      node.id = this.consume("string", "id value").value;
      return;
    }
    if (key.value === "path") {
      node.paths = this.parseStringOrList();
      return;
    }
    if (artefactKeys.includes(key.value as never)) {
      node.artefacts[key.value] = this.parseStringOrList();
      return;
    }
    throw new ParseError(`Unknown attribute '${key.value}'`, { source: this.source, position: key.position });
  }

  private parseStringOrList(): string[] {
    if (this.match("bracketOpen")) {
      const values: string[] = [];
      while (!this.check("bracketClose") && !this.check("eof")) {
        values.push(this.consume("string", "string list item").value);
        if (!this.match("comma") && !this.check("bracketClose")) {
          throw this.expected(", or ]");
        }
      }
      this.consume("bracketClose", "]");
      return values;
    }
    return [this.consume("string", "string value").value];
  }

  private parseEdge(): Edge {
    const source = this.consume("identifier", "edge source");
    this.consume("arrow", "->");
    const target = this.consume("identifier", "edge target");
    const description = this.consume("string", "edge description");
    return {
      source: source.value,
      target: target.value,
      description: description.value,
      position: source.position,
    };
  }

  private peek(): Token {
    return this.tokens[this.index]!;
  }

  private check(type: TokenType): boolean {
    return this.peek().type === type;
  }

  private checkNext(type: TokenType): boolean {
    return this.tokens[this.index + 1]?.type === type;
  }

  private advance(): Token {
    return this.tokens[this.index++]!;
  }

  private match(type: TokenType): boolean {
    if (!this.check(type)) return false;
    this.advance();
    return true;
  }

  private consume(type: TokenType, label: string): Token {
    if (this.check(type)) return this.advance();
    throw this.expected(label);
  }

  private expected(label: string): ParseError {
    const token = this.peek();
    return new ParseError(`Expected ${label}, found ${token.type}${token.value ? ` '${token.value}'` : ""}`, {
      source: this.source,
      position: token.position,
    });
  }
}

File: src/dsl/parser.rs
//! Recursive descent parser for the Phase 1 Cairn DSL grammar.

use std::{fs, path::Path};

use super::{
    ast::{Ast, Edge, Field, Node, NodeKind, Span},
    error::{ParseError, ParseErrorKind},
    lexer::{Token, TokenKind, tokenize},
};

/// Parses a DSL file from disk.
///
/// # Errors
///
/// Returns an I/O-backed parse error when the file cannot be read, or a
/// source-positioned parse error when the DSL is malformed.
pub fn parse_file(path: impl AsRef<Path>) -> Result<Ast, ParseError> {
    let path = path.as_ref();
    let source = fs::read_to_string(path).map_err(|error| {
        let span = Span::point(path.display().to_string(), 1, 1);
        ParseError {
            code: "CAIRN_IO_READ_DSL",
            message: format!("failed to read DSL: {error}").into_boxed_str(),
            span: Box::new(span),
            kind: Box::new(ParseErrorKind::UnexpectedToken {
                expected: "readable DSL file".to_owned(),
                encountered: "I/O error".to_owned(),
            }),
        }
    })?;
    parse_str(&path.display().to_string(), &source)
}

/// Parses DSL source.
///
/// # Errors
///
/// Returns a source-positioned parse error when the DSL is malformed.
pub fn parse_str(file: &str, source: &str) -> Result<Ast, ParseError> {
    Parser {
        tokens: tokenize(file, source)?,
        index: 0,
    }
    .parse()
}

struct Parser {
    tokens: Vec<Token>,
    index: usize,
}

impl Parser {
    fn parse(&mut self) -> Result<Ast, ParseError> {
        let mut nodes = Vec::new();
        let mut edges = Vec::new();
        while !self.at_eof() {
            if self.peek_node_kind().is_some() {
                nodes.push(self.node()?);
            } else {
                edges.push(self.edge()?);
            }
        }
        Ok(Ast { nodes, edges })
    }

    fn node(&mut self) -> Result<Node, ParseError> {
        let kind_token = self.next().clone();
        let kind = node_kind(&kind_token).ok_or_else(|| {
            ParseError::unexpected(
                kind_token.span.clone(),
                "node declaration",
                describe(&kind_token.kind),
            )
        })?;
        let name = self.expect_word("node name")?;
        let description = self.expect_string("node description")?;
        self.expect_word_exact("id")?;
        let id = self.expect_string("node id")?;
        let mut tags = Vec::new();
        while let TokenKind::Tag(tag) = self.peek().kind.clone() {
            self.next();
            tags.push(tag);
        }
        self.expect_kind(&TokenKind::OpenBrace, "`{`")?;
        let mut paths = Vec::new();
        let mut owns_files = false;
        let mut contracts = Vec::new();
        let mut raw_fields = Vec::new();
        let mut children = Vec::new();
        while !self.consume_kind(&TokenKind::CloseBrace) {
            if self.at_eof() {
                return Err(ParseError::unexpected(
                    self.peek().span.clone(),
                    "`}`",
                    "end of input",
                ));
            }
            if self.peek_node_kind().is_some() {
                children.push(self.node()?);
                continue;
            }
            let field_span = self.peek().span.clone();
            let field_name = self.expect_word("field name")?;
            match field_name.as_str() {
                "path" => paths.extend(self.field_values()?),
                "contract" => contracts.extend(self.field_values()?),
                "owns-files" => {
                    self.expect_kind(&TokenKind::Colon, "`:`")?;
                    let value = self.expect_word("boolean")?;
                    owns_files = value == "true";
                }
                _ => {
                    let values = self.field_values()?;
                    raw_fields.push(Field {
                        name: field_name,
                        values,
                        span: field_span,
                    });
                }
            }
        }
        Ok(Node {
            kind,
            name,
            description,
            id,
            tags,
            paths,
            owns_files,
            contracts,
            raw_fields,
            children,
            span: kind_token.span,
        })
    }

    fn edge(&mut self) -> Result<Edge, ParseError> {
        let span = self.peek().span.clone();
        let from = self.expect_word("edge source id")?;
        self.expect_kind(&TokenKind::Arrow, "`->`")?;
        let to = self.expect_word("edge target id")?;
        let description = self.expect_string("edge description")?;
        Ok(Edge {
            from,
            to,
            description,
            span,
        })
    }

    fn field_values(&mut self) -> Result<Vec<String>, ParseError> {
        if self.consume_kind(&TokenKind::OpenBracket) {
            let mut values = Vec::new();
            while !self.consume_kind(&TokenKind::CloseBracket) {
                values.push(self.expect_string("list item string")?);
                let _ = self.consume_kind(&TokenKind::Comma);
            }
            Ok(values)
        } else {
            Ok(vec![self.expect_string("field value string")?])
        }
    }

    fn expect_word_exact(&mut self, word: &str) -> Result<(), ParseError> {
        let token = self.next().clone();
        match token.kind {
            TokenKind::Word(value) if value == word => Ok(()),
            kind => Err(ParseError::unexpected(token.span, word, describe(&kind))),
        }
    }

    fn expect_word(&mut self, expected: &str) -> Result<String, ParseError> {
        let token = self.next().clone();
        match token.kind {
            TokenKind::Word(value) => Ok(value),
            kind => Err(ParseError::unexpected(
                token.span,
                expected,
                describe(&kind),
            )),
        }
    }

    fn expect_string(&mut self, expected: &str) -> Result<String, ParseError> {
        let token = self.next().clone();
        match token.kind {
            TokenKind::String(value) => Ok(value),
            kind => Err(ParseError::unexpected(
                token.span,
                expected,
                describe(&kind),
            )),
        }
    }

    fn expect_kind(&mut self, kind: &TokenKind, expected: &str) -> Result<(), ParseError> {
        let token = self.next().clone();
        if &token.kind == kind {
            Ok(())
        } else {
            Err(ParseError::unexpected(
                token.span,
                expected,
                describe(&token.kind),
            ))
        }
    }

    fn consume_kind(&mut self, kind: &TokenKind) -> bool {
        if &self.peek().kind == kind {
            self.next();
            true
        } else {
            false
        }
    }

    fn peek_node_kind(&self) -> Option<NodeKind> {
        node_kind(self.peek())
    }

    fn at_eof(&self) -> bool {
        self.peek().kind == TokenKind::Eof
    }

    fn peek(&self) -> &Token {
        &self.tokens[self.index]
    }

    fn next(&mut self) -> &Token {
        let token = &self.tokens[self.index];
        if token.kind != TokenKind::Eof {
            self.index += 1;
        }
        token
    }
}

fn node_kind(token: &Token) -> Option<NodeKind> {
    match &token.kind {
        TokenKind::Word(value) if value == "System" => Some(NodeKind::System),
        TokenKind::Word(value) if value == "Container" => Some(NodeKind::Container),
        TokenKind::Word(value) if value == "Module" => Some(NodeKind::Module),
        TokenKind::Word(value) if value == "Actor" => Some(NodeKind::Actor),
        _ => None,
    }
}

fn describe(kind: &TokenKind) -> String {
    match kind {
        TokenKind::Word(value) => format!("word `{value}`"),
        TokenKind::String(value) => format!("string `{value}`"),
        TokenKind::Tag(value) => format!("tag `@{value}`"),
        TokenKind::OpenBrace => "`{`".to_owned(),
        TokenKind::CloseBrace => "`}`".to_owned(),
        TokenKind::OpenBracket => "`[`".to_owned(),
        TokenKind::CloseBracket => "`]`".to_owned(),
        TokenKind::Comma => "`,`".to_owned(),
        TokenKind::Colon => "`:`".to_owned(),
        TokenKind::Arrow => "`->`".to_owned(),
        TokenKind::Eof => "end of input".to_owned(),
    }
}
